### ISSUE TITLE
Add parameter value hint in the json

### ISSUE THREAD

**Issue by XapaJIaMnu** [hplt-project/OpusCleaner#1]
https://github.com/jelmervdl/empty-train/blob/5edf6e20b7fe381ca87d4abe9aa4fcf1985b63ef/main.py#L31
We should have optionally a list of possible values, or something like a datatype hint (eg int)

**Comment by jelmervdl** [comment]
Yep! That's why it exists, but still empty because I didn't know yet how to describe them.

It might be useful to just add a bunch of the filters we already have, and see what their parameters are in practice. Then we can see which types we'd need, and whether the current implementation that just uses shell variables is flexible enough.

**Comment by jelmervdl** [comment]
Moving this here from #2 so I can merge that without hiding this:

For context:

```py
FILTER_PARAM = { "filter_alpha_ratio": {
        "command": "filters/alpha_ratio.py --src-lang $LANG1 --trg-lang $LANG2 --ratio-words-src $SRCWORDRAT\
         --ratio-words-trg $TRGWORDRAT --ratio-alpha-src $SRCALPHARAT --ratio-alpha-trg $TRGALPHARAT",
        "parameters": {
            "LANG1": {"type": "str", "allowed_values": list(CHARS.keys())},
            "LANG2": {"type": "str", "allowed_values": list(CHARS.keys()) + [None]},
            "SRCWORDRAT": {"type": "float", "default": 0.4},
            "TRGWORDRAT": {"type": "float", "default": 0.4},
            "SRCALPHARAT": {"type": "float", "default": 0.5},
            "TRGALPHARAT": {"type": "float", "default": 0.5}
        }
    }
}
```

I like these options, but I'm conflicted about how you do `allowed_values` here. I want to support this, it's pretty essential. Just don't know how yet.

I want to avoid parsing external python files just to figure out all available filters. Like, I'm not sure how that would work out when you'd have more complex filters like bicleaner which might be installed in their on virtualenv with their own dependencies. It's pretty easy to make them callable from outside. Hell, if all else fails, you could wrap them in a shell script that loads the proper environment etc.

Most of the parameters can also just be described in a json file. That would not lock us into python, and versioning json schemas is a lot easier than maintaining different versions of some python API. However, options like the `available_options` key, which are dynamic here, wouldn't work.

One work-around for that could be to allow a `$(...)` sort of syntax that could be used to ask the filter to print all available options. You could then do something like:

```json
{
  "allowed_values": "$(filters/alpha_ratio.py --help | sed -n 's/^ *--src-lang {\\(.*\\)}/\\1/p' | tr ',' '\\n')"
}
```

… but looking at that example, it looks pretty unwieldy.

Another one, if most filter tools use argparse anyway, would be to just parse the output of `filter.py --help`. It looks pretty standardised and parsable. Only downside I see right now is that there is no way to differentiate between str and float arguments.

Two other problems are: boolean and optional parameters.

Flags, or boolean arguments like `--debug` won't work with the simple variable replacement method used now.

Same for optional parameters. How do I specify no value for `--src-lang` so it falls back to the default? Leaving `$LANG1` blank won't remove the `--src-lang` bit from the command.

**Comment by jelmervdl** [comment]
Maybe the dynamic `available_options` is looking for a complex solution to a luxury problem (the alternative: just copy/paste the supported values into the parameter description json, and if you update one, you update the other.)

Possible solutions for the boolean flag and the optional parameters:

Using [the `${parameter+alt_value}, ${parameter:+alt_value}` variable substitution option](https://tldp.org/LDP/abs/html/parameter-substitution.html#PARAMALTV) in the shell, you can do stuff like `${DEBUG:+--debug}` for adding `--debug` only if `$DEBUG` is not empty. And `${LANG1:+--src-lang $LANG1}` for only printing the `--src-lang` bit if `$LANG1` is not empty.

My argument for bash-isms here: `command` is already passed to the shell, and that does perform variable substitution. So this is practically implemented for free. Also, people who write these scripts are probably pretty familiar with Bash, or can find plenty of examples of how to do this stuff in bash.

**Comment by XapaJIaMnu** [comment]
Hi,

I generally agree on the keeping it simple. Here are several proposals:
I could add a simple command such as `--dump_parameters`, that would produce a a `json` with all the necessary values, maybe in a separate directory and then the main script scans that directory for `json` files and auto adds filters.

For filters such as bash based or perl ones, that require mostly static options, one could manually generate the jsons. That way you move away from the python dependency and the dynamic generation.

One step further could be to even add a hook that checks for new filters on push, and tries to call each of them with `--dump_parameters` to auto generate the jsons, but I think that might be overdoing it.

**Comment by jelmervdl** [comment]
Hmm, that does sound nice and simple. Maybe for the Python case we can write a little function that hooks into argparse to get that json automatically. For Python at least that would save you from writing the parameter schema twice. For other binaries, they can implement it themselves.

I'd think it would be as follows: add a second type of json filter file that just points to the command that generates the parameter schema json. The code that gathers all available filters runs that command on start-up. That way you can manually add filter json pointers to include filters installed all over your system. And when one of those filters gets updated, you automatically get the latest options even though the filter program itself doesn't need to know about where emtpy-train is installed and where it should write parameter json files.

**Comment by jelmervdl** [comment]
If we're autogenerating the JSON anyway, let's just treat options with default values (where the default is not `None`) as normal options. I don't think it is bad to make the defaults explicit.

**Comment by jelmervdl** [comment]
I've taken your script, and turned it into a stand-alone script that dumps the json for python script that uses argparse, e.g. `./dump-parameter-schema.py filters/max-length.py`.

It monkey patches the `argparse` module, and then executes the script you pass it as if that were the script that was actually being executed. At the moment that then tries to parse the arguments (signalling all parameters have been defined) it dumps the schema and quits.

```py
#!/usr/bin/env python3
import sys
import argparse
import json
from typing import Any, cast

script_path = sys.argv[1]

# Remove the script name from the argument list
sys.argv[1:] = sys.argv[2:]


def _argparse_parse_args(self: argparse.ArgumentParser, args=None, namespace=None):
	"""Creates a json string from the argparser instance"""
	json_out: dict[str,Any] = {}
	json_out["description"] = self.description
	
	# non-simple type so it doesn't end up in json
	SWITCH = object() 
	SUBSTITUTE = object()

	# We need to skip [0], as this is the prepended `--help`
	param_dict: dict[str,dict] = {}
	for argument in self._actions[1:]:
		current_str = {
			SWITCH: argument.option_strings[0], #TODO prefer long names?
			SUBSTITUTE: argument.option_strings[0].replace('-','').upper(),
			"type": cast(type, argument.type).__name__ if type(argument.type) == type else "bool",
			"required": argument.required,
			"default": argument.default,
			"help": argument.help
		}

		if argument.choices is not None:
				current_str["allowed_values"] = argument.choices
		
		# Add to the parameter dict
		param_dict[current_str[SUBSTITUTE]] = current_str

	json_out["parameters"] = param_dict

	json_out["command"] = script_path
	for _, value in param_dict.items():
			if value["type"] == "bool":
				json_out["command"] += " ${" + value[SUBSTITUTE] + ":+" + value[SWITCH] + "}"
			else:
				json_out["command"] += " " + value[SWITCH] + " $" + value[SUBSTITUTE]

	json.dump(json_out, sys.stdout, indent=4, skipkeys=True)
	sys.exit(0)


argparse.ArgumentParser.parse_args = _argparse_parse_args

with open(script_path) as fh:
	exec(fh.read())
```
